lib/Dialect/OM/Transforms/LinkModules.cpp

```diff
@@ -0,0 +1,294 @@
+                                      ArrayRef<ClassLike> classes) {
+  bool existExternalClass = false;
+  size_t countDefinition = 0;
+  ClassOp classOp;
+
+  for (auto op : classes) {
+    if (isa<ClassExternOp>(op))
+      existExternalClass = true;
+    else {
+      classOp = cast<ClassOp>(op);
+      ++countDefinition;
+    }
+  }
+
+  // There must be exactly one definition if the symbol was referred by an
+  // external class.
+  if (existExternalClass && countDefinition != 1) {
+    SmallVector<Location> classExternLocs;
+    SmallVector<Location> classLocs;
+    for (auto op : classes)
+      (isa<ClassExternOp>(op) ? classExternLocs : classLocs)
+          .push_back(op.getLoc());
+
+    auto diag = emitError(classExternLocs.front())
+                << "class " << name << " is declared as an external class but "
+                << (countDefinition == 0 ? "there is no definition"
+                                         : "there are multiple definitions");
+    for (auto loc : ArrayRef(classExternLocs).drop_front())
+      diag.attachNote(loc) << "class " << name << " is declared here as well";
+
+    if (countDefinition != 0) {
+      // There are multiple definitions.
+      for (auto loc : classLocs)
+        diag.attachNote(loc) << "class " << name << " is defined here";
+    }
+    return failure();
+  }
+
+  if (!existExternalClass)
+    return countDefinition != 1;
+
+  assert(classOp && countDefinition == 1);
+
+  // Raise errors if linked external modules are not compatible with the
+  // definition.
+  auto emitError = [&](Operation *op) {
+    auto diag = op->emitError()
+                << "failed to link class " << name
+                << " since declaration doesn't match the definition: ";
+    diag.attachNote(classOp.getLoc()) << "definition is here";
+    return diag;
+  };
+
+  llvm::MapVector<StringAttr, Type> classFields;
+  for (auto fieldOp : classOp.getOps<om::ClassFieldOp>())
+    classFields.insert({fieldOp.getNameAttr(), fieldOp.getType()});
+
+  for (auto op : classes) {
+    if (op == classOp)
+      continue;
+
+    if (classOp.getBodyBlock()->getNumArguments() !=
+        op.getBodyBlock()->getNumArguments())
+      return emitError(op) << "the number of arguments is not equal, "
+                           << classOp.getBodyBlock()->getNumArguments()
+                           << " vs " << op.getBodyBlock()->getNumArguments();
+    unsigned index = 0;
+    for (auto [l, r] : llvm::zip(classOp.getBodyBlock()->getArgumentTypes(),
+                                 op.getBodyBlock()->getArgumentTypes())) {
+      if (l != r)
+        return emitError(op) << index << "-th argument type is not equal, " << l
+                             << " vs " << r;
+      index++;
+    }
+    // Check declared fields.
+    llvm::DenseSet<StringAttr> declaredFields;
+    for (auto fieldOp : op.getBodyBlock()->getOps<om::ClassExternFieldOp>()) {
+      auto it = classFields.find(fieldOp.getNameAttr());
```

Comment:
Is it required to store `classFields` in a separate data structure? Is it possible for this to use `lookupSymbol` or is it necessary to keep a separate data structure?

Reply:
Ah I see. I wasn't thinking about the use-case to visit all defined symbols, and yeah, seems like that API doesn't exist :/